src/libmatches_macro/lib.rs

```diff
@@ -0,0 +1,29 @@
+#![no_core]
+#![feature(no_core)]
+#![feature(staged_api)]
+#![doc(test(no_crate_inject))]
+
+/// Returns whether the given expression matches (any of) the given pattern(s).
+///
+/// # Examples
+///
+/// ```
+/// #![feature(matches_macro)]
+/// use std::macros::matches;
+///
+/// let foo = 'f';
+/// assert!(matches!(foo, 'A'..='Z' | 'a'..='z'));
+///
+/// let bar = Some(4);
+/// assert!(matches!(bar, Some(x) if x > 2));
+/// ```
+#[macro_export]
+#[unstable(feature = "matches_macro", issue = "0")]
+macro_rules! matches {
+    ($expression:expr, $( $pattern:pat )|+ $( if $guard: expr )?) => {
```

Comment:
I would soon like to try a crater run for having the `pat` matcher allow `$pat_0 | ... | $pat_n` and this would be a roadblock towards that.

Reply:
Hm, how so? Seems like if we implemented that expansion then we could drop the repetition here, right?

Reply:
Ah yes that seems true.

Reply:
Since this is in the standard library you could change `$( $pattern:pat )|+` to `$pattern:pat` when landing the language change, right?